File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(combination, currSum, start):
            if currSum == target:
                result.append(combination[:])
                return
            if currSum > target:
                return

            for i in range(start, len(candidates)):
                combination.append(candidates[i])
                backtrack(combination, currSum + candidates[i], i)
                combination.pop()

        candidates.sort()
        result = []
        backtrack([], 0, 0)
        return result

File: 0039-combination-sum/0039-combination-sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
    const result: number[][] = [];

    function backtrack(index: number, target: number, combination: number[]) {
        if (target === 0) {
            result.push([...combination]);
            return;
        }

        for (let i = index; i < candidates.length; i++) {
            if (candidates[i] <= target) {
                combination.push(candidates[i]);
                backtrack(i, target - candidates[i], combination);
                combination.pop();
            }
        }
    }

    candidates.sort((a, b) => a - b);

    backtrack(0, target, []);

    return result;
};